### PR TITLE
Add cache for capabilities client

### PR DESCRIPTION
**Pull request by mhenschke-atl**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -15,6 +15,7 @@
         <bitbucket.version>5.5.9</bitbucket.version>
         <cloverVersion>4.3.1</cloverVersion>
         <disableTestInjection>false</disableTestInjection>
+        <google.guava.libversion>29.0-jre</google.guava.libversion>
```
Comment (dkjellin) [inline pom.xml:18]: part of the ticket was to work out which was the best caching solution, can you elaborate on why we picked Guava? We _know_ guava will give us headaches if we're not extremely careful, so I'm keen to hear why we didn't go with the cache built into okhttp for example

Comment (mhenschke-atl) [reply pom.xml:18]: I tried importing it, and given it looks like we're actually already using Guava in a few places, it seemed like it would be okay to use. With the one failing integration test excepted, I'm not sure what headaches it's caused. That said, are there other issues with this solution we can't see in this PR? If so, happy to close and use a different solution (okhttp sounds good)

**Comment by mhenschke-atl** [comment]
By removing the direct dependency on Guava and relying on Jenkin's version as I've done in `mh/cache-capabilities-jenkins-guava` (https://mvnrepository.com/artifact/com.google.guava/guava/11.0.1) this passes integration tests, including on Java 11. So the question changes from whether or not we could to whether or not we *should*.
Jenkins also makes use of guava caches so I think we're at least as compatible with Java 11 as Jenkins is. An alternative, if we don't want to use this 8-year-old library is https://square.github.io/okhttp/4.x/okhttp/okhttp3/-cache/. This unfortunately has to live in the network layer rather than just on top of the client factory, but again it's not a major change so this is a potential alternative if we prefer.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/supply/BitbucketCapabilitiesSupplier.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    @Override
+    public AtlassianServerCapabilities get() {
+        HttpUrl url =
+                bitbucketRequestExecutor.getBaseUrl().newBuilder()
+                        .addPathSegment("rest")
+                        .addPathSegment("capabilities")
+                        .build();
+        return bitbucketRequestExecutor.makeGetRequest(url, AtlassianServerCapabilities.class).getBody();
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/supply/BitbucketCapabilitiesSupplier.java:23]: My guava cache knowledge is a bit rusty (which arguably is an advantage given the, uhm, vintage, version used).
Anyway, what about edge cases here? What happens in case of an exception?

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImplTest.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    @InjectMocks
+    private BitbucketCapabilitiesClientImpl capabilitiesClient;
+
+    @Test
+    public void testGetServerCapabilitiesNoCache() {
+        Mockito.when(capabilitiesSupplier.get()).thenReturn(newCapabilities);
+        assertEquals(newCapabilities, capabilitiesClient.getServerCapabilities());
+        Mockito.verify(capabilitiesSupplier).get();
+    }
+
```
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImplTest.java:29]: I think a test showing how it works when an exception is thrown from the supplier may be in order.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/supply/BitbucketCapabilitiesSupplier.java (hunk 2 above)
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/supply/BitbucketCapabilitiesSupplier.java:23]: That's a good question- will add a test for it. There aren't any checked exceptions so I'm going to assume the behaviour we want is the same as it was without the cache- any runtime errors bubble up to the consumer where they can handle them?

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImplTest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    @Mock
+    private BitbucketCapabilitiesSupplier capabilitiesSupplier;
+    @Mock
+    private AtlassianServerCapabilities newCapabilities, cachedCapabilities;
+    @InjectMocks
+    private BitbucketCapabilitiesClientImpl capabilitiesClient;
+
+    @Test
+    public void testGetServerCapabilitiesNoCache() {
+        Mockito.when(capabilitiesSupplier.get()).thenReturn(newCapabilities);
```
Comment (alibk0rd) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImplTest.java:25]: super nit (for consistency with other places): static import for Mockito methods like `when()` and `verify()`, etc.

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImplTest.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    @Test
+    public void testGetServerCapabilitiesWithCache() {
+        Mockito.when(capabilitiesSupplier.get()).thenReturn(cachedCapabilities);
+        capabilitiesClient.getServerCapabilities();
+        Mockito.verify(capabilitiesSupplier).get();
+
+        assertEquals(cachedCapabilities, capabilitiesClient.getServerCapabilities());
+        Mockito.verifyNoMoreInteractions(capabilitiesSupplier.get());
```
Comment (alibk0rd) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImplTest.java:37]: I may be slightly misreading this, but I'm guessing this line was meant for verifying no more calls to `capabilitiesSupplier.get()`?
I _think_ what this verifies now (as opposed to `verifyNoMoreInteractions(capabilitiesSupplier)` for example), is that it says "verify no more interactions with the object/entity returned from `capabilitiesSupplier.get()`", i.e. no more interactions with `cachedCapabilities`.
There's no easy/graceful way to verify no more interactions with a specific method of a mock in Mockito that I know of unfortunately.

Good news is I just checked locally and `Mockito.verifyNoMoreInteractions(capabilitiesSupplier);` works in this case.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImpl.java [hunk 6]
```diff
@@ -1,35 +1,36 @@
...
 
 public class BitbucketCapabilitiesClientImpl implements BitbucketCapabilitiesClient {
 
     private final BitbucketRequestExecutor bitbucketRequestExecutor;
+    private final Supplier<AtlassianServerCapabilities> capabilitiesCache;
 
-    BitbucketCapabilitiesClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor) {
+    BitbucketCapabilitiesClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor, BitbucketCapabilitiesSupplier supplier) {
         this.bitbucketRequestExecutor = bitbucketRequestExecutor;
+        capabilitiesCache = Suppliers.memoizeWithExpiration(supplier, 60L, TimeUnit.MINUTES);
```
Comment (alibk0rd) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImpl.java:23]: nit: `private static final` field for timeout?

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImplTest.java (hunk 5 above)
Comment (mhenschke-atl) [reply src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketCapabilitiesClientImplTest.java:37]: Ah good catch! Thank you
